// react-rtc-real/dist/ReactRTC/functions/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createPayload = exports.createMessage = exports.generateRoomKey = exports.buildServers = void 0;

var buildServers = function buildServers(serverURLs) {
  if (!serverURLs) return;
  var urls = serverURLs.map(function (serverURL) {
    return {
      urls: serverURL
    };
  }); // eslint-disable-next-line consistent-return

  return urls;
};

exports.buildServers = buildServers;

var replaceCharacter = function replaceCharacter(_char) {
  // eslint-disable-next-line no-mixed-operators
  // eslint-disable-next-line no-bitwise
  // eslint-disable-next-line prettier/prettier
  return (_char ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> _char / 4).toString(16);
};

var generateRoomKey = function generateRoomKey() {
  var baseString = [1e7] + -1e3 + -4e3 + -8e3 + -1e11;
  return baseString.replace(/[018]/g, replaceCharacter);
};

exports.generateRoomKey = generateRoomKey;

var createMessage = function createMessage(type, payload) {
  return {
    type: type,
    payload: payload
  };
};

exports.createMessage = createMessage;

var createPayload = function createPayload(roomKey, socketID) {
  var message = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  return {
    roomKey: roomKey,
    socketID: socketID,
    message: message
  };
};

exports.createPayload = createPayload;
